KEP-1977: Add KEP for ContainerNotifier

xing-yang opened the pull request:
Enhancement issue: https://github.com/kubernetes/enhancements/issues/1977

xing-yang commented:
CC @yuxiangqian

xing-yang commented:
/assign @thockin @saad-ali @liggitt 
/assgin @sjenning @derekwaynecarr @dchen1107

xing-yang commented:
/assgin @sjenning @derekwaynecarr @dchen1107

thockin commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,770 @@
+
+NotificationStatus represents the current state of a Notification. It has a list of PodNotificationStatuses.
+
+A PodNotificationStatus represents the current state of a Notification for a specific pod. A PodNotificationStatus contains ContainerNotificationStatuses of all the containers which have the corresponding ContainerNotifier specified. PodNotificationStatus is created when a Notifier starts to run in a pod.
```
This is going to be a problem.  Imagine a notifier that selects 1000 pods.  That's not implausible.  Even 5000 is believable.  That leaves you just a few hundred bytes per pod for status, before it starts failing (too big).  Worse, it will fail in non-deterministic ways in practice.

I think we have a few choices here, and we should seek a pattern precedent (@liggitt for consideration)

1) Notification.status is a "summary".  The real status is stored in another new resource `NotificationResult` which is selected by something like "notification.kubernetes.io/uid=<notification-uid>".  Each NotificationResult holds the status for one single pod-notification.  Need to define the lifecycle management of this so we don't leave them around for too long.

2) Notification.status is a "summary".  The real status is stored in each pod, e.g. in `status.notifications[]` which would be a list of "recent" notifications and the per-container statuses

3) other ideas?

TL;DR: we can't have (practically) unbounded lists in a single resource.

xing-yang replied:
Ok, I'll wait for more suggestions.

xing-yang replied:
Are there any existing examples that I could follow?

thockin replied:
EndpointSlice is similar - I am not sure there's a closer example.

liggitt replied:
> NotificationStatus represents the current state of a Notification. It has a list of PodNotificationStatuses.

when we talked about this in doc form, I thought notification was scoped to a single pod in order to allow [notifying in waves](https://docs.google.com/document/d/1SWSlZoxY5zFjBKFKaATP07s3q02UenSp8R9-yRkCcwg/edit?disco=AAAAJ1RQggU) and [avoid unbounded status accumulation](https://docs.google.com/document/d/1SWSlZoxY5zFjBKFKaATP07s3q02UenSp8R9-yRkCcwg/edit?disco=AAAAKEpj35c) ... did that approach get revised since then?

xing-yang replied:
So that means we can't have a PodSelector in the NotificationSpec.  It will just be a PodName instead.

xing-yang replied:
@thockin would you consider having one to one mapping between a Notification and a Pod?  That would simplify status update.

krmayankk replied:
Just read the discussion in the doc, I think the notification per pod makes most sense and seems safer. Also the idea of building on top using a notification controller for safe rollout of notification is even better

thockin replied:
If we go to 1:1, do we need a higher-level API that selects pods and expands a notification for each of them?or should we add that to kubectl?  or should we just say "someone else's problem"?

If we go 1:1 does a subresource make more sense?  It gives even finer RBAC control, but we need the result to be a new resource (for status).  E.g. POST to /pods/foobar/notify -> creates a PodNotifier; kubelet watches PodNotifiers.  Any precedent for that?  Any need to get finer-grained than per-namespace?

krmayankk replied:
I like the subresource idea which i was hinting here https://github.com/kubernetes/enhancements/pull/1995/files#r493239279 . I guess i don't follow why it should result in a new PodNotifier being created? For e.g. does the POST on a log or the exec sub resource result in a new object ? The main reason i like this is its less number of new objects to manager , less strain on etcd, since the notifications are an array. It does make the RBAC more fine grained. Thinking out loud, if i am not mistaken, you can say a service account can or cannot exec to any pods in a namespace, So similar should be possible with this notify subresource

thockin commented on keps/sig-node/1977-container-notifier/README.md:
````diff
@@ -0,0 +1,770 @@
+All containers in a selected pod which have a "ContainerNotifier" defined in their spec with the "ContainerNotifierName" will have their corresponding "ContainerNotifierAction" executed. Moreover, there is no guarantee of ordering on execution.
+
+```
+type Notification struct {
````
We need to define what happens in this case:

time 0: pods A, B, C are running, labelled foo=bar
time 1: create Notification selecting foo=bar
time 2: all notifiers in A, B, C are done
time 3: pod D is created, labelled foo=bar
*** Do we run the notifier on D?  If so, for how long does that apply?  If not, how do we prevent it?

xing-yang replied:
We have said that Kubelet (container notifier controller) will not do retries.  So if that is the case, Kubelet (container notifier controller) will only process the pods matching the label when the Notification object is first created?  If so, Pod D will not be processed?

krmayankk replied:
Is this a namespaced object ?  Once its created, it should apply to everything  until the api object is deleted . 
Is this more of an imperative api or we want it to be declarative (i guess the latter )?
i am thinking what does the declaration mean ?
- notify all containers that match this selector at this time t
- let there be notification for all pods matching this selector as long this api exists
- i want all containers of the pods matching a selector to be quiesced (this seems more natural)
its a bit confusing to think since the lifecycle hooks are tied to the lifecycle of the container/pod and are ok to be imperative, but this api is not tied to the lifecycle but more of a command to the containers.

thockin replied:
@krmayankk is on the issue.

This is a pseudo-imperative operation.

@xing-yang 


> Kubelet (container notifier controller) will only process the pods matching the label when the Notification object is first created

That seems reasonable at first blush, but what is the mechanism for that?  Is the rule "only applies to pods whose `metadata.creationTimestamp` is less than the Notification's `metadata.creationTimestamp` ?  If so, let's document it.  Also - less-than or less-than-or-equal?

Does this satisfy the higher-order controller's needs?  Is it plausible that a quiesce notification could be created, a snapshot started, then a new pod shows up which is not quiesced?  Even if Notifications applied to it, that would be async, so it still seems bad?

yuxiangqian replied:
@thockin using CreationTimestamp can certainly be one option to rule out newly created pods, it does not cover the case when a pod is evicted/died during hook execution. From quiescing requirements perspective, its critical to have a point-in-time "fixed" scope for snapshotting, thus if a pod died during the notification, we might want to fail the the whole notification.
we have a couple of other potential options here:
1. make notification to pod one-to-one by replacing podSelector with a pod definition. higher level controller like application snapshot controller then can have a pre-selected list and create one notification to each desired pod.
Advantage:

- simpler API
- ordering of notification is possible and more convenient to achieve for higher level controllers
disadvantages:
 a. probably does not fit well with the signal case where multiple pods need to be notified without a higher level controller

2. introducing a list of resolved pod in Notification.Spec which will be immutable once resolved. Pod selection is done prior to any real execution of notification. And that list would be the source of truth for the scope of execution.
WDYT?

thockin replied:
This is an interesting problem.  If a pod dies mid-snapshot, do you fail the whole snapshot?  If a new pod a starts mid-snapshot (scale-up?) does it need to quiesce?  Maybe that just "doesn't happen" ?

Your (1) happens to ALSO solve the status cardinality problem, so is attractive in that regard.  It means the scale of things kubelet has to watch just got much bigger, though, unless we set labels to scope it (e.g. set the node-name as a label, let kubelet watch only that).

But it is ugly for humans.

Another approach would be to define Notifications as applying to all pods.  In the example above it *would* apply to D.  And E, F, G etc until the Notification is removed.  All the controller (or kubelet) could guarantee would be that every Notification that selects a given pod would be attempted once, and if there were multiple notifications, they would be attempted in some deterministic order (e.g. creationTimestamp or RV).

So to expand the example:

t0:  pods A, B, C are running, labelled foo=bar
t1:  create Notification N1 selecting foo=bar
t2:  N1 notifiers in A, B, C are done
t3:  pod D is created, labelled foo=bar
t4:  N1 notifier in D is done
t5:  create Notification N2 selecting foo=bar
t6:  N2 notifiers in A, B, C, D are done
t7:  pod E is created, labelled foo=bar
t8:  N1 notifier in E is done
t9:  N2 notifier in E is done
t10: delete Notification N1
t11: pod F is created, labelled foo=bar
t12: N2 notifier in F is done  ## note N1 was not executed
t13: delete Notification N2

This seems maybe workable, but I need you to look at it from your use-cases.  Bear in mind that rapid sequences can be observed out of order by watchers, so if your plan would be to delete N1 then create N2 you need some time in between to know that all kubelets have observed the deletion (or ensure that it is "safe" to trigger N1 on new pods).  Similarly, if you want to clean up N1 and N2 you *must* delete them in order with sufficient time between to ensure no out-of-order deletion problems.

We don't have a "serializing barrier" operation. 

Regarding "ugly for humans", maybe we need both?  Individual pod-notifications (and status) which controllers can orchestrate PLUS a selector-notification which a built-in controller would expand into pod-notifications?  This represents a broader API change, but is the more flexible model.

Thoughts?

Regarding your (2) - does that satisfy requirements?  Let me see if I get it right:

t0:  pods A, B, C are running, labelled foo=bar
t1:  create Notification N1 selecting foo=bar
t2:  a controller evaluates foo=bar one time and notes the result (A, B, C)
t3:  N1 notifiers in A, B, C are done
t4:  pod D is created, labelled foo=bar
t5:  N1 notifier in D is NOT executed

Is that right?  How does that map to the use-cases?

krmayankk replied:
Applying to all pods for as long as the notification exists makes more sense from an api point of view. Also including the pod definition would be messy.

krmayankk replied:
> @krmayankk is on the issue.
> 
??

> This is a pseudo-imperative operation.
> 
Is that ok to have ? Do we need to update any apiconvention guidelines on when a pseudo imperative api makes more sense ?

xing-yang replied:
For the quiesce use case, a higher level backup controller needs to decide what pods are included. When a backup is started, the backup controller needs to figure out all the pods that need to get the quiesce command.  It can retrieve this information using a PodSelector.  After that it is a fixed list of Pods that need to be quiesced for this particular backup.  Therefore, we can’t have PodSelector in Notification.  It should be either a List of Pods that is immutable or a Pod name for a single pod, basically #2 and #1 suggested by Xiangqian.

If we decide to have a PodList in NotificationSpec, the steps would be like this:

1. An external Backup controller determines what pods need to receive the notification by using a PodSelector (this depends on backup controller logic).
2. Backup controller creates a Notification object with a PodList with the selected pods.

This way we have a concrete list of Pods. If quiesce fails for any pod/container, it should be treated as failure for the whole quiesce operation.  Typically the backup software gives user an option to complete a backup that is not quiesced successfully or fail it if any quiesce action fails. If user chooses to do a backup even if quiesce fails, the backup controller will trigger a un-quiesce command to undo anything that is already quiesced but skip doing quiesce for the remaining pod/container.  Then the backup controller will take a snapshot and mark that snapshot as not-quiesced.  User can decide what to do with a snapshot that is not quiesced.

thockin replied:
> we can’t have PodSelector in Notification

I'm not sure I buy that.  This API needs to be general enough to use by humans.  We need to be clear about what happens if a new pod is created that satisfies the criteria.  Don't over-focus on just quiesce

thockin commented on keps/sig-node/1977-container-notifier/README.md:
````diff
@@ -0,0 +1,770 @@
+}
+```
+
+### Phase 3 API Additions
````
I think we should say something like "Phase 3 is more loosely decided than phases 1 and 2, and if it is needed, we will have to revisit the design details".  There are just too many variables to see the needs at this point (IMO).

As such, I am not aggressively API reviewing it :)

xing-yang replied:
Ok

thockin commented on keps/sig-node/1977-container-notifier/README.md:
````diff
@@ -0,0 +1,770 @@
+}
+```
+
+To support "SIGHUP" directly, we can either convert it to the command "kill -SIGHUP" and send it the same way as other commands, or we can translate that into a call to the docker client method `ContainerKill`.
````
you can't assume a container has a `kill` binary to execute

xing-yang replied:
Removed wording using the command.

thockin commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,770 @@
+
+### Alpha -> Beta Graduation
+
+* One release has been waited to allow for feedback from users.
```
Are we willing to go Beta with the phase 1 design?  I don't think so, am I wrong?  I think you should talk about phases here.  If we decide we need phase 3, will we go beta without it?  What if we go beta and THEN decide we need phase 3, does that become alpha and we release independently?  Or do we wait for it to "catch up" and GA together.

xing-yang replied:
We'll need phase 2 before going Beta.  Regarding phase 3, we should discuss and make a decision after phase 2 is done?

thockin commented on keps/sig-node/1977-container-notifier/kep.yaml:
```diff
@@ -0,0 +1,50 @@
+  - "@liggit"
+  - "@saad-ali"
+  - "@sjenning"
+approvers:
```
That's a lot of approvers - do we need them all?  I feel like Jordan OR me, Dawn OR Derek..

xing-yang replied:
Sure, we'll trim down the approvers list.

xing-yang commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,770 @@
+
+            // PodSelector specifies a label query over a set of pods.
+            // +optional
+	PodSelector *metav1.LabelSelector
```
I assume you meant list "PodSelector" as the first parameter and "ContainerNotifierName" second in NotificationSpec?

krmayankk commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,773 @@
+
+NotificationStatus represents the current state of a Notification. It has a list of PodNotificationStatuses.
+
+A PodNotificationStatus represents the current state of a Notification for a specific pod. A PodNotificationStatus contains ContainerNotificationStatuses of all the containers which have the corresponding ContainerNotifier specified. PodNotificationStatus is created when a Notifier starts to run in a pod.
```
why a specific pod? this should contain the status of notification for all pods that match the selector right ?

xing-yang replied:
For a particular PodNotificationStatus, it represents status for a particular pod.  Yes, all pods matching the selector will be selected.

krmayankk commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,773 @@
+    // +optional
+    // define constants for signals?
+    // validate the signals are valid?  windows?
+++    Signal string
```
Do we need to start with signal in phase 1 and then add exec later if that is necessary ?

xing-yang replied:
Support for signal requires changes to CRI. That means it can only happen in Kubelet.  Since we are starting with a separate controller, we can't start with signal.

Also support for exec is an important use case (the quiesce use case for application consistent snapshot). So support for exec should be added in phase 1.

thockin replied:
We should not go GA (and maybe not beta) without phase 2, but it is phased for a reason.

krmayankk commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,773 @@
+
+### Kubelet Impact in Phase 2 and Beyond
+
+When moving the logic from the container notifier controller to Kubelet, there will be impact on Kubelet.
```
its not clear from this proposal why the controller is starting with a separate controller(also not clear if its separate or part of one of the existing controllers) and then later moving to kubelet

xing-yang replied:
This will start as a separate controller in a new repo, and later on the logic will be moved to kubelet.
This is because there are concerns from sig-node on potential impact on kubelet.

thockin replied:
An abundance of caution

krmayankk commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,773 @@
+
+### Phase 1
+
+Phase 1 API definition will be added to Kubernetes. Implementation in this phase will not happen in Kubelet. Instead, it will be implemented in a single *trusted* controller which acts on Notifications via exec (with the goal being to move that into Kubelet in the next steps). In this phase, only exec will be included and signal will not be included as that involves changes to CRI.
```
so a separate external controller running per node (daemon set) ? In order to run the exec commands , will it exec into the pods ?

xing-yang replied:
Yes.

thockin replied:
not needed per-node, just per-cluster

xing-yang commented:
CC @andrewsykim

thockin commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,773 @@
+}
+
+// ContainerNotifier describes a command to run in the container.
+type ContainerNotifierAction struct {
```
I was proposing that we get rid of this layer of nesting:

```
Pod {
  Spec: {
    Containers: [{
      Notifiers: [{
        Name: foo
        TimeoutSeconds: 10
        Action: {   # note no "handler"
          Exec: {
            Command: bar
          }
        }
      }]
    }]
  }
}
```

sjenning replied:
yes, agreed

krmayankk replied:
+1

thockin replied:
ping

thockin commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,773 @@
+           // Names are label-style, with an optional prefix. Names without the prefix
+           // such as `quiesce` are reserved for Kubernetes-defined "well-known" names.
+           // Names with a prefix such as `example.com/label-style` are custom names.
+           ContainerNotifierName string
```
minor: This is a needlessly long name.  Just `Notifier` or `NotifierName` perhaps?

krmayankk replied:
It would be good to think about who can send notifications ? What permissions they need on a pod ?. Since the controller is doing exec, does it mean as long as exec is opened for the controller, anyone can create and run the commands using controller as a proxy?

Another idea is should we make notifier a subresource ?

thockin replied:
RBAC governs who can create Notifications in a namepace - isn't that enough?

krmayankk replied:
see here https://github.com/kubernetes/enhancements/pull/1995/files#r555640516

sjenning commented:
I would like to see example pod specs (at least the container blocks) for the examples and use cases mentioned.

I understand I should be able to think about what it would look like from the API definition, but just to make it clearer.

It exposes issues like what @thockin noticed [here](https://github.com/kubernetes/enhancements/pull/1995#discussion_r492865819)

yuxiangqian commented on keps/sig-node/1977-container-notifier/kep.yaml:
```diff
@@ -0,0 +1,48 @@
+# The milestone at which this feature was, or is targeted to be, at each stage.
+milestone:
+  alpha: "v1.20"
+  beta: "v1.21"
```
I think this is too aggressive given that the amount of work.

xing-yang replied:
These milestones need to be updated.

thockin reviewed (commented):
Corner cases are so much fun

krmayankk commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,773 @@
+    // which selects a pod and matches the Notifier name, the Notifier will run in the
+    // container.
+    // +optional
+    Notifiers []ContainerNotifier
```
do you think we could have named the lifecycle hooks and reused them ?

krmayankk replied:
Do we need this at pod level, to mean it applies to all containers ?

thockin replied:
I don't think reusing hook definitions is worth the effort - early versions of this had a new resource just for that and I found it quite cumbersome.  It's part of the pod, just like a probe.

thockin replied:
Re pod-level - I think that's an optimization (for laziness) which we can explore if it comes up.  I doubt it.

xing-yang commented:
CC @ashish-amarnath

thockin reviewed (commented):
Given the open comments should we discuss again?  Is this stalled on a decision or on hands-on-keyboards?

xing-yang commented:
> Given the open comments should we discuss again? Is this stalled on a decision or on hands-on-keyboards?

@thockin Yes, I think we should discuss again as there are a few open issues.  I'll try to set up a meeting.

yuxiangqian commented:
> Given the open comments should we discuss again? Is this stalled on a decision or on hands-on-keyboards?
>
@thockin there are couple of other things(VolumeSnapshot GA, KubeCon US etc) going on concurrently for both @xing-yang and me, and we should have more spins after this week. I am reworking some pieces of this KEP for a bit to make it more clear, and we will schedule a discussion to talk about the Notification piece. Thanks for your patience.

thockin commented:
NP - I have a fair bit of free time for the next few weeks, so I am circling back to all the KEPs I am tracking

thockin commented:
1.21 KEP freeze is in O(days) - if this wants review before then, I need updates.

yuxiangqian commented:
Apologies for the late update. We have revisited some of the designs in this KEP to address some of the major concerns raised in the previous version(details in this [document](https://docs.google.com/document/d/1lPXtF58qpdrpR6vsRREGTzhPRkHOc1qMbcPrnjfxrgU/edit#)).

Major changes:
1. Resolve the concerns around potential scalability issues when a Notification targets at a large number of Pods(~Ks).
2. Shrink the scope in phase 1 to reduce implementation complexity and simpler PodNotification API.
3. Resolve too-deeply nested structure concern.

yuxiangqian commented:
/test pull-enhancements-verify

shawn-hurley commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,877 @@
+    // Upon creation of a PodNotification, Container(s) in the specified
+    // Pod with a matching notifier name will be executed.
+    // +optional
+    Notifiers []ContainerNotifier
```
Is this mutable? Should someone be able to add a notifier to a pod after it is running I guess is the question that I am asking?

yuxiangqian replied:
"Containers" field in Pod can not be updated that is my understanding.
[ref](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#podspec-v1-core)
IMO, it also seems to be reasonable to make the "notifiers" immutable such that only a restricted/predefined set of notifiers can be triggered. Are there any specific cases you see the needs of dynamically adding/removing notifiers from a specific container?

shawn-hurley replied:
I was wondering if it would be possible to do that, I can't think of a good use case thanks!

shawn-hurley commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,877 @@
+    // Number of seconds after which the notifier will time out.
+    // Default to 1 second. Minimum value is 1.
+    // +optional
+    TimeoutSeconds int32
```
Should there be a Handler that is run on timeout?

yuxiangqian replied:
How to handle timeout/errors is indeed an area that the current API proposal intentionally tries to simply and transfer some of the complexity to users. Could you give some concrete examples? One thing that bothers me for introducing  error/timeout handlers is that another "timeout" might be needed for the error/timeout handler itself. It would also complicate the API/implementation quite a bit. WDYT?

shawn-hurley replied:
I agree that it will complicate things, what I was worried about is if I quiesce an application but  it times out, to bring the application back up to running, we would want to run an unquiesce operation? 

I guess this could be solved by the notifier having a graceful termination with a shutdown signal being sent on timeout? Wondering if it should be apart of the enhancement?

yuxiangqian replied:
the main issue with introducing a graceful termination in the API is that it does not fundamentally solve the problem. It's possible the graceful termination mechanism itself is not graceful enough :). Meaning, it might end up as a failure. It seems to be reasonable to me that Applications/controllers understands the semantic of returned errors of a notification, and handle it appropriately (i.e., keep on retrying an un-quiesce hook until it succeeds).

thockin commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,877 @@
+
+When moving the logic from the PodNotification controller to Kubelet, we foresee the following changes to happen in Kubelet:
+
+- Kubelet will watch PodNotification resources cluster-wide.
```
It only needs to watch notifications for pods on this node.  That's a potentially important optimization we should plan for,

yuxiangqian replied:
agreed. Updated this point.

yuxiangqian commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,877 @@
+// More error types could be added, e.g., Forbidden, Unauthorized, etc.
+const (
+    // Pod does not exist
+    NonExistingPod PodNotificationErrorType = "NonExistingPod"
```
ack

yuxiangqian commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,877 @@
+    // Note: Only Pods in the ‘Running’ phase are eligible for notification.
+    // Ref: kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase
+    // Required.
+    PodSelector metav1.LabelSelector
```
ack. done

yuxiangqian commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,877 @@
+
+    // The name of the ContainerNotifier to trigger.
+    // Name is a DNS_LABEL.
+    Notifier string
```
removed this comment.

yuxiangqian commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,877 @@
+    // AllPods means all Pods matches the specified label selector in Notification
+    // will be notified. If this policy is specified, CompleteTime field of the
+    // NotificationStatus will not be set.
+    AllPods PodSelectionPolicy = "AllPods"
```
For the "AllPod" one, we can simply remove this type, "PodSelectionPolicy" is just an optional field in spec, we can treat the default behavior to "AllPod".
For the "PreExistingPods" option, it's really to support the case for application consistency snapshot.

I see there's a lot of confusions and concerns over this model, and there is indeed another alternative which is acceptable to me:

1. Get rid of the PodSelectionPolicy completely from Notification, and the behavior will be always try to notify any newly created Pod which matches the label selector until the Notification resource is deleted.
2. For those upper level controllers which rely on completion of PodNotifications, i.e., quiesce a pod before taking a snapshot of it's volumes, they can always use the `PodNotification` type to get finer grained control. (Having the upper level controller/end user to handle the complexity, and leave the API simple).

I personally are fine for both models, and if there are not other cases, I would actually prefer the alternative to have a simpler API.

WDYT?

@xing-yang any comments?

yuxiangqian replied:
also I would like to point out this is a phase 2 item, and we can always adjust the design should there be more use cases or new requirements

thockin replied:
> For the "PreExistingPods" option, it's really to support the case for application consistency snapshot.

I'm still not convinced that would be correct.  If a new pod started during a snapshot (buggy controller?) wouldn't you want to notify that one?  That raises a fun corner case - suppose I deliver the notification to pod P, and then there was an OOM kill during the snapshot, causing pod P to restart.  We would want to re-deliver the notification, right?  What ensures that, and how do I report it?

yuxiangqian replied:
IMO, these cases should ALL be treated as a failure of the Notification. The definition of PreExistingPods is to only cover those ones are running prior to the creation of a Notification.
If PreExistingPods is specified,
1. Any existing Pod crashes/restarts fails the Notification
2. Newly created Pod will *not* be notified

If re-delivery is needed, a user can always elect to use "AllPods"?

yuxiangqian commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,877 @@
+    // Note that this field will never be set if "AllPods" is specified as
+    // pod selection policy.
+    // +optional
+    CompleteTime *metav1.Time
```
see comments above for the alternative, if we achieve an agreement there for the proposed alternative, then we can remove this field.

I assume you meant having a GoalCount in `NotificationSpec`? That has the risk racing in the following situation:
1. prior to create notification, pod count = X
2. create a notification with label selector and goalcount = X
3. a pod died, or a new pod has been created, now the pod count = Y
4. during execution, X != Y, which ones to run the command if Y > X? or Y < X, should it fail?

thockin replied:
I meant under status - the controller would update the goal count if new pods start that match the selector.  Then a client could look at the number of completions and the goal count.  When that "settles" (which is really context dependent) then we know it's "done enough".  That way it applies in all modes (all pods or pre-existing).

yuxiangqian commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,877 @@
+
+    // Reason is a brief CamelCase string that describes the failure and it’s
+    // meant for machine parsing and tidy display in the CLI.
+    Reason string
```
This type used to be  a Condition type, and changed to be an error. removed reason.

thockin commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -246,48 +246,51 @@ type PodNotificationStatus struct {
+    // of execution is unknown(i.e., still in execution).
     // +optional
-    Succeeded *metav1.Time
+    Succeeded *bool
```
I think this is better served by an enum `result` (or something like that) with values for success and failure, so we can later add other things.

Thoughts?

yuxiangqian replied:
make sense, will update

thockin commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -312,11 +311,11 @@ type PodNotificationErrorType string
-    // Pod does not exist
-    NonExistingPod PodNotificationErrorType = "NonExistingPod"
+    // Pod not found error
+    PodNotFound PodNotificationErrorType = "PodNotificationErrorPodNotFound"
```
You changed this the opposite of what I meant :)

`PodNotifictionErrorPodNotFound PodNotificationErrorType = "PodNotFound"`

yuxiangqian replied:
Ack

thockin commented:
I am pretty close to happy with this.

thockin reviewed (commented):
Thanks!

/lgtm
/approve

ruiwen-zhao commented on keps/sig-node/1995-container-notifier/README.md:
```diff
@@ -0,0 +1,899 @@
+    PodName string
+
+
+    // The name of the ContainerNotifier to trigger.
```
PodNotificationSpec only has PodName and ContainerNotifier name (which is only container-level unique). So if a pod has two containers, each of which has a ContainerNotifier with the same name, how does PodNotificationSpec distinguish between them?

xing-yang replied:
If two containers have the same ContainerNotifier, then the same command will be run in both containers; If the intention is to run different commands in two different containers, each should use a different ContainerNotifier name.

ruiwen-zhao replied:
Got it, as long as this is intended we should be good. Actually you already mentioned that "All containers in the specified `Pod` which have the requested "ContainerNotifier" defined in their spec will have their corresponding "ContainerNotifierHandler" executed".

thockin replied:
It notifies all containers with that named notifier

yuxiangqian commented on keps/sig-node/1995-container-notifier/README.md:
```diff
@@ -0,0 +1,899 @@
+    PodNotificationErrorPodNotFound PodNotificationErrorType = "PodNotFound"
+
+    // The execution of a handler timed out.
+    PodNotificationErrorHandlerTimeout PodNotificationErrorType = "HandlerTimeout"
```
yes, it will be surfaced both in PodNotificationStatus aggregated field "State" as well as detailed in "Containers" which users can tell in which container the notifier failed.

thockin commented:
/lgtm

xing-yang commented:
KEP number in kep.yaml should be the enhancement issue number 1977.  Fixed it now.

sjenning commented:
/remove-lifecycle stale

Phase 1 looks acceptable to me.

The kubelet needs no change; only the pod spec.  The kubelet does not need to be notification aware in phase 1.

Completion of phase 1 will allow us to prove out this idea and root out potential issues with the design, both API and performance/scale, before lumping the controller logic into the kubelet where it will be harder to debug and isolate as a source of scaling issues.

dchen1107 commented:
@sjenning thanks for taking another look on this. Like what I mentioned at last week's SIG Node, the new proposal narrowed down the scope a lot based on our offline review meetings and comments. Due to the overwhelm of the reviewers last cycle, we postponed this feature, and hopefully we can finish the phase 1 for 1.23 this cycle.

derekwaynecarr commented:
I had a number of questions/clarifications I hope we can get through in SIG Node call, otherwise I will paste here right after.

sjenning commented:
Derek and I talked about this.  Some thoughts.

Point in time nature of Notifications. Notification created at time X might behave differently from one created at time X+1, due to the PreExisting SelectorPolicy and Pod creation times i.e. if the PodNotification is created before the Pod, the notification fails. PodNotifications are a one shot thing.

Notification controller watches all pods for CREATE (UPDATE?) to see if the selector matches and then creates PodNotifications on container start? container ready? pod ready?  Do pods that have deletionTimestamp set get notifications?

If the pod is deleted, are all PodNotifications that reference the pod deleted too?  Is the ownerRef for the PodNotification the pod? or the Notification?

Number of PodNotification resources in the cluster would be the sum of all matching pods over all Notifications, a potentially very large number.  How are these cleaned up?  Notification controller (i.e. kubelet in the future) would have to WATCH PodNotifications cluster-wide right?  I don't see a way to apiserver-side filter the watch to only PodNotifcations referencing a pod a particular kubelet is running.  Thus for each change to a PodNotification cluster-wide, a watch event is sent to each kubelet.

QPS for the controller updating all these PodNotification resources will need to be high.  Stress on container runtime and apiserver.  A potential scaling foot-gun.  We have had issues with kubelet QPS in the past delaying pod status updates and leading to all manner of pain.

We have a number of container types including init and ephemeral containers that also use the Container type.  Do notifications work for them?  Doesn't seem to make sense for those container types.
Point in time nature of Notifications. Notification created at time X might behave differently from one created at time X+1, due to the PreExisting SelectorPolicy and Pod creation times i.e. if the PodNotification is created before the Pod, the notification fails. PodNotifications are a one shot thing.

Notification controller watches all pods for CREATE (UPDATE?) to see if the selector matches and then creates PodNotifications on container start? container ready? pod ready?  Do pods that have deletionTimestamp set get notifications?

If the pod is deleted, are all PodNotifications that reference the pod deleted too?  Is the ownerRef for the PodNotification the pod? or the Notification?

Number of PodNotification resources in the cluster would be the sum of all matching pods over all Notifications, a potentially very large number.  How are these cleaned up?  Notification controller (i.e. kubelet in the future) would have to WATCH PodNotifications cluster-wide right?  I don't see a way to apiserver-side filter the watch to only PodNotifcations referencing a pod a particular kubelet is running.  Thus for each change to a PodNotification cluster-wide, a watch event is sent to each kubelet.

QPS for the controller updating all these PodNotification resources will need to be high.  Stress on container runtime and apiserver.  A potential scaling foot-gun.  We have had issues with kubelet QPS in the past delaying pod status updates and leading to all manner of pain.

We have a number of container types including init and ephemeral containers that also use the Container type.  Do notifications work for them?  Doesn't seem to make sense for those container types.

derekwaynecarr commented:
> I don't see a way to apiserver-side filter the watch to only PodNotifcations referencing a pod a particular kubelet is running. Thus for each change to a PodNotification cluster-wide, a watch event is sent to each kubelet.

I think for this item if we can integrate with the `NodeAuthorizer` admission plugin, it would be good.

derekwaynecarr commented:
To summarize:

I think this feature needs a garbage collection strategy so we understand the anticipated life time of a PodNotification resource as both a provider/consumer.  I would think its reasonable to say that a Pod can have X active PodNotification resources per pod.  This helps us budget QPS, plan for handler costs, etc.  Completed PodNotification resources must be garbage collected in order to avoid denial of service for both the kubelet (memory growth) and kube-api (similar).  TTL on the resource is one way to do that.

Exec/probe costs are really expensive in practice and a key input in understanding node sizing, so having a fixed budget to plan for is helpful.  For awareness, pods whose liveness or readiness probes are done with too great of frequency or make heavy use of exec dominate cpu resource usage by the kubelet and runtime.  This is why having a limit on the number of concurrent notifications at admission time is helpful.

Adding ContainerNotificationStatus to include ContainerID so it maps to what is shown in PodStatus would let us know if the value is unique across time and space.

When in the pod lifecycle is the pod accepting signals, is a signal ever needed before success of particular probes (liveness/readiness/etc.)?  If not, should we scope it to a particular well-defined phase in the Pod lifecycle.

Is there a reason to not explicitly exclude init containers or ephemeral containers as notification targets?

If a pod is undergoing termination, is a signal still sent?

If a notification is created for a terminal pod, who updates the status?

If a notification is created that references a notifier that does not exist in the pod spec, what happens?  Does that fail at validation time or runtime?

Should the notification reference a pod by UUID and not just name so it is also bound in time and space?

smarterclayton commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,899 @@
+In phase 1, propose to introduce several API changes:
+- Adding an optional field `Notifiers` which is a list of `ContainerNotifier`s into `Container`.
+- Adding an inlined type `ContainerNotifierHandler` which defines a command.
+- Adding a core API type `PodNotification` which defines request type to trigger execution of `ContainerNotifier`(s) in a `Pod`.
```
I don’t see a defense/justification for this object in the examples of this KEP (or alternatives) vs simply having a spec generation driven interaction in the pod itself for phase one.  If it’s been discussed and rejected, I’d like it added to alternatived (so i can see the critique).  If it hasn’t been discussed, I’d prefer it to the addition of a new object and controller in phase one that seem like they are “controller for controllers sake”.

Roughly, id expect pod.spec.containers.notifier to have a blank generation field that defaults to spec generation on edit of the notifier struct, or is 0 on create.  And then status to have “notifiers[].generation” that indicates if the kubelet executed the signal.  Clients clear spec generation to trigger the signal and wait for status generation to match to know when the kubelet successfully “leveled” the notifier (delivered it).

That seems a more useful primitive than the hypothetical object and controller, can be done in phase one so external controllers can invoke signals, and doesn’t require the extra alpha baggage until later (bulk notification can be done as a separate orthogonal phase around generation driven signaling).

thockin replied:
To be honest, this design was trying to minimize the number of moving parts within Pod, but if folks want to explore that as an alternative...

Pod doesn't seem to use `generation` so first we enable that.  Or use RV ?

So to trigger a notifier, we'd write 0 to the pod...notifier generation, which is roughly equivalent to creating a PodNotifier as specced today.  Right?

That does get rid of PodNotifier type in exchange for more moving parts in Pod itself.  In particular if means that any controller which converts a higher-level Notifier (eg, notify all pods under a selector) needs RBAC to mutate Pod.  That seems strictly worse to me, unless we add a subresource for notifications.

Do you perceive that as better?  I don't hate it, and it eliminates a whole category of problems.

We still need a controller.  Two in fact - one to actuate the notifications (eventually kubelet) and one to take higher level "notify by selector" and convert that to N discrete notifications (or we push that up onto all consumers of notifiers).  As you say, that's not phase 1 necessary.

yuxiangqian commented:
> To summarize:
> 
> I think this feature needs a garbage collection strategy so we understand the anticipated life time of a PodNotification resource as both a provider/consumer. I would think its reasonable to say that a Pod can have X active PodNotification resources per pod. This helps us budget QPS, plan for handler costs, etc. Completed PodNotification resources must be garbage collected in order to avoid denial of service for both the kubelet (memory growth) and kube-api (similar). TTL on the resource is one way to do that.
> 
Added a TTLSecondsAfterCompletion field to PodNotificationSpec and NotificationSpec
> Exec/probe costs are really expensive in practice and a key input in understanding node sizing, so having a fixed budget to plan for is helpful. For awareness, pods whose liveness or readiness probes are done with too great of frequency or make heavy use of exec dominate cpu resource usage by the kubelet and runtime. This is why having a limit on the number of concurrent notifications at admission time is helpful.
> 
Will address this in a separate commit.
> Adding ContainerNotificationStatus to include ContainerID so it maps to what is shown in PodStatus would let us know if the value is unique across time and space.
> 
Added
> When in the pod lifecycle is the pod accepting signals, is a signal ever needed before success of particular probes (liveness/readiness/etc.)? If not, should we scope it to a particular well-defined phase in the Pod lifecycle.
> 
Add clarification to the API. A `PodNotification` should fail immediately if a `Pod` has `PodPhase` other than `PodPending` or `PodReady`, in case of `PodPending`, the PodNotification will hold until the Pod is ready.
> Is there a reason to not explicitly exclude init containers or ephemeral containers as notification targets?
Good point, made it clear in API that InitContaines and EphemeralContainers will be ignored.
> 
> If a pod is undergoing termination, is a signal still sent?
Clarified in API, only `PodReady` and `PodPending` will be considered. For a pod undergoing termination, I am thinking about scenarios like a container can not be terminated successfully (stuck in some loop for example), and a signal is needed to sent to the container to force kill. Does that sound like a reasonable use case? Another aspect is, only containers still in "Running" state will have the notification executed.
> 
> If a notification is created for a terminal pod, who updates the status?
In phase one, that would be the **trusted** controller, and phase 2 will be Kubelet.
> 
> If a notification is created that references a notifier that does not exist in the pod spec, what happens? Does that fail at validation time or runtime?
I assume you mean PodNotification here, I am thinking an admission control to this kinda of validation to avoid the PodNotification from creation. Will address in a separate commit.
> 
> Should the notification reference a pod by UUID and not just name so it is also bound in time and space?
Good Point, added UID into PodNotificationStatus

xing-yang commented:
/assign @johnbelamaric 
PRR is ready for review.

thockin commented:
Given Clayton's comment, it seems unlikely that this will be resolved in time for 1.23 - I am hopeful, but trying to be realisitic.

johnbelamaric commented:
Perhaps this has been discussed, but I have to ask why this needs to be a core API.

It would seem the original motivation - run a quiesce command in each Pod - could be handled via an agent container living in the Pod that is watching a CR. Time may be better spent designing an out-of-tree solution based on a CR and a general agent that can run in a container in the Pod and act locally on those pod containers. Unless there is some reason that we need a privileged host-level process like `kubelet` to take some action, I can't see a justification for adding this level of complexity for something that can be solved without it.

xing-yang commented:
/retest

thockin commented:
I think I disagree with John on this one, but I admit I am pretty close to it, which colors my perspective.  It might be worth trying.  For example, using a sidecar to enact your own notifications might be possible - how streamlined could it get?

johnbelamaric commented:
This is on my PRR list but I don't see any updates since pre-1.23. Assume this is a no-go for 1.24?

chenxuc commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,947 @@
+
+## Motivation
+
+Being able to take application consistent snapshots/backups is a requirement to protect stateful workloads in Kubernetes. It's required by many applications to temporally quiesce themselves before taking snapshots of their PersistentVolume(s) and unquiesce afterwards to ensure that an application consistent snapshot/backup can be creatd. A mechanism to send a quiesce/unquiesce command to `Pod`(s) ...
```
`snapshot/backup can be creatd` change to:
`snapshot/backup can be created`

chenxuc commented on keps/sig-node/1977-container-notifier/README.md:
```diff
@@ -0,0 +1,947 @@
+
+This example involves a signal so it will be in phase 2 and beyond.
+
+For example, a user wants to send SIGHUB to a container in a `Pod`. This signal is defined in the `ContainterNotificationHandler` inside the `Container`.
```
`wants to send SIGHUB` change to:
`wants to send SIGHUP`

chenxuc reviewed (commented):
Found some typos. Thanks.

xing-yang commented:
/remove-lifecycle stale

thockin commented:
I don't suppose this is going anywhere in 1.25?

xing-yang commented:
> I don't suppose this is going anywhere in 1.25?

@thockin Unfortunately no.  We still have those unresolved comments.  Xiangqian and I need to sync up about them.

thockin commented:
I don't suppose anyone is on this for 1.26?

xing-yang commented:
> I don't suppose anyone is on this for 1.26?

Thanks @thockin for the ping.  Unfortunately we are not ready to target 1.26.

sftim commented on keps/sig-node/1977-container-notifier/kep.yaml:
```diff
@@ -0,0 +1,48 @@
+owning-sig: sig-node
+participating-sigs:
+  - sig-storage
+status: implementable
```
To help this move forward, how about we try to merge it with
```suggestion
status: provisional
```

and then refine from there?

xing-yang replied:
Sure.  We could try that.

thockin commented:
That's a good idea - @xing-yang any objections?

xing-yang commented:
> That's a good idea - @xing-yang any objections?

@thockin Sure, let's try that.  Thanks.

thockin commented:
Can you make the proposed change and drop the PRR file?  Then we can merge as provisional.

xing-yang commented:
> Can you make the proposed change and drop the PRR file? Then we can merge as provisional.
Hi @thockin, I've dropped the PRR file and changed the status to "provisional".

thockin reviewed (commented):
Thanks!

/lgtm
/approve

thockin commented:
@dchen1107 can you approve this as "provisional" ?  Nobody is working on it but I don't want to lose state.

thockin commented:
I'm just going to close this and if anyone ever gets time, we can reopen
